software-mansion/react-native-reanimated
# Fix "Reading from `_value` directly is only possible on the UI runtime." on fast refresh during library development

## Summary

This PR fixes one of the most annoying errors in Reanimated that often appears during development of the library as well as sometimes in client apps.

<img width="505" alt="Zrzut ekranu 2023-11-29 o 21 50 06" src="https://github.com/software-mansion/react-native-reanimated/assets/20516055/89fc45aa-68d6-4f1d-99be-0959ee64df16">

## Investigation

When mutable is created, we call `registerShareableMapping(mutable, handle)` to store the handle in  `_shareableCache`.

When cloning a worklet, for each mutable in its closure, in `makeShareableCloneRecursive` we obtain the handle using `_shareableCache.get` method and skip the execution of the rest of the function body.

During a fast refresh, React holds the same instance of the mutable (that's guaranteed by `useRef`) but for some reason `_shareableCache` is regenerated and becomes an empty WeakMap.

Because of it, `makeShareableCloneRecursive` can't find the handle in `_shareableCache` and tries to clone the mutable as if it was a regular JS object.

We use `Object.entries` to iterate over the keys which throws an error when `_value` property is accessed.

## Solution

The proposed solution is to simply move the definition of `_shareableCache` to a separate file. During a fast refresh, the contents of this file usually doesn't change (unless you modify this exact file on purpose) so the cache is not recreated which solves the issue.

## Remarks

I also replaced all calls to `_shareableCache.set` with `registerShareableMapping` which does exactly the same.

## Test plan

1. Open `PlatformChecker.ts`
2. Add `return true;` in the implementation of `isReducedMotion()`
3. Save the file. This should trigger a fast refresh.
4. Observe the error "[Reanimated] Reading from _value directly is only possible on the UI runtime."
5. Apply the changes from this PR, repeat steps 1-4, the error should not appear again.
